File: frontend/src/services/followUpService.js
/**
 * JobTrack Follow-Up REST API Service
 * Handles all communication with the Spring Boot backend (/api/v1/follow-ups and /api/v1/jobs/{jobId}/follow-ups) with JWT Authentication.
 */

const TOKEN_KEY = 'jobtrack_jwt_token';

/**
 * Get HTTP headers with Authorization Bearer token if present.
 */
function getAuthHeaders(customHeaders = {}) {
  const token = localStorage.getItem(TOKEN_KEY);
  const headers = {
    'Accept': 'application/json',
    ...customHeaders,
  };

  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }

  return headers;
}

/**
 * Helper to process HTTP responses and parse JSON or error payloads.
 */
async function handleResponse(response) {
  const contentType = response.headers.get('content-type');
  const isJson = contentType && contentType.includes('application/json');
  const body = isJson ? await response.json() : null;

  if (!response.ok) {
    if (response.status === 401) {
      console.warn('Received 401 Unauthorized from backend API.');
    }

    const errorMessage =
      (body && (body.message || (body.fieldErrors && Object.values(body.fieldErrors).join(', ')))) ||
      `Request failed with status ${response.status} (${response.statusText})`;

    const error = new Error(errorMessage);
    error.status = response.status;
    error.data = body;
    throw error;
  }

  return body;
}

export const followUpService = {
  /**
   * Fetch all follow-up reminders for the authenticated user.
   * @param {boolean|null} completed Optional filter for completion status (true / false / null)
   * @returns {Promise<Array>} Array of follow-up response objects
   */
  async getAllFollowUps(completed = null) {
    let url = '/api/v1/follow-ups';
    if (completed !== null && completed !== undefined) {
      url += `?completed=${Boolean(completed)}`;
    }

    const response = await fetch(url, {
      method: 'GET',
      headers: getAuthHeaders(),
    });
    const result = await handleResponse(response);
    return result?.data || [];
  },

  /**
   * Create a new follow-up reminder under a job application.
   * @param {number|string} jobId
   * @param {Object} followUpData
   * @returns {Promise<Object>} Created follow-up object
   */
  async createFollowUp(jobId, followUpData) {
    const response = await fetch(`/api/v1/jobs/${jobId}/follow-ups`, {
      method: 'POST',
      headers: getAuthHeaders({
        'Content-Type': 'application/json',
      }),
      body: JSON.stringify(followUpData),
    });
    const result = await handleResponse(response);
    return result?.data;
  },

  /**
   * Toggle the isCompleted status of a follow-up reminder.
   * @param {number|string} id
   * @returns {Promise<Object>} Updated follow-up object
   */
  async toggleFollowUp(id) {
    const response = await fetch(`/api/v1/follow-ups/${id}/toggle`, {
      method: 'PATCH',
      headers: getAuthHeaders(),
    });
    const result = await handleResponse(response);
    return result?.data;
  },
};

File: backend/src/main/java/com/jobtrack/service/FollowUpService.java
package com.jobtrack.service;

import com.jobtrack.dto.request.FollowUpRequest;
import com.jobtrack.dto.response.FollowUpResponse;

import java.util.List;

/**
 * Service interface defining business operations for Follow-Up reminders with multi-user data isolation.
 */
public interface FollowUpService {

    /**
     * Create a new follow-up reminder under a specific job application belonging to the authenticated user.
     *
     * @param jobId   the parent job application ID
     * @param request the validated follow-up payload
     * @param userId  the authenticated user ID
     * @return the created follow-up response
     */
    FollowUpResponse createFollowUp(Long jobId, FollowUpRequest request, Long userId);

    /**
     * Retrieve all follow-ups for the authenticated user, optionally filtered by completion status.
     *
     * @param userId    the authenticated user ID
     * @param completed optional boolean filter (true = completed only, false = pending only, null = all)
     * @return list of follow-up responses
     */
    List<FollowUpResponse> getAllFollowUps(Long userId, Boolean completed);

    /**
     * Retrieve a specific follow-up by its ID and authenticated user ID.
     *
     * @param id     the follow-up ID
     * @param userId the authenticated user ID
     * @return the follow-up response
     */
    FollowUpResponse getFollowUpById(Long id, Long userId);

    /**
     * Toggle the completion status (isCompleted) of a follow-up reminder.
     *
     * @param id     the follow-up ID to toggle
     * @param userId the authenticated user ID
     * @return the updated follow-up response
     */
    FollowUpResponse toggleFollowUpCompletion(Long id, Long userId);

    /**
     * Update an existing follow-up reminder.
     *
     * @param id      the follow-up ID to update
     * @param request the updated follow-up payload
     * @param userId  the authenticated user ID
     * @return the updated follow-up response
     */
    FollowUpResponse updateFollowUp(Long id, FollowUpRequest request, Long userId);

    /**
     * Delete a follow-up reminder.
     *
     * @param id     the follow-up ID to delete
     * @param userId the authenticated user ID
     */
    void deleteFollowUp(Long id, Long userId);

    // Overloads for non-user-scoped operations
    FollowUpResponse createFollowUp(Long jobId, FollowUpRequest request);
    List<FollowUpResponse> getAllFollowUps(Boolean completed);
    FollowUpResponse getFollowUpById(Long id);
    FollowUpResponse toggleFollowUpCompletion(Long id);
    FollowUpResponse updateFollowUp(Long id, FollowUpRequest request);
    void deleteFollowUp(Long id);
}
